### PR TITLE
(Re)load content on main objects pages

### PR DESCRIPTION
**Pull request by robbevp**
# Description
This PR adds all functionality to (re)load the content on the main object pages (Album, Artist, Genre, Label).
This means:
* Going directly to a specifc page (`app/albums/1`) will get content and functionality almost instantly.
* Content is much more likely to stay up to date.

There are two drawbacks to this:
* We introduce a lot more API calls. Every page for a main object will do 2 -3 API calls. I would try this and see if it creates issues/too much traffic. If so, we can check `object.loaded` before calling the api and only call the API if the content is older than some timeperiod.
* There is a higher chance that related content is missing. This is something that should/will be added in a future PR.

## Scopes
To accomplish this, I've added a few scopes classes to our api interface.
These classes allow the user to easily create a scope without bothering with the implementation on the api-side.
A scope can be:
* created on a single line `new AlbumsScope.label(id)`
* chained for more complex queries `new AlbumsScope.label(id).artist(id).filter(string)`
* created and then modified:
```
const scope = new Albumscope()
scope.label(id)
scope.artist(id)
scope.finalQuery
```

The scope class throws errors if we pass it the wrong kind of parameter. This errors are not translated since the end-user should not encounter these. They are only their to notify the develop if they made an error when building scopes.

If we use a scope, we skip `removeOld` after fetching all data. In the future, we might want to allow that method check whether any objects that fit the scope should be removed, but IMHO that is not really relevant. (If a user visits the page for an object that doesn't exist on the api anymore, they will get an error and go back to the previous page).

# How Has This Been Tested?

- [x] Local tests to see if everything works
- [x] A quick test with our production api, as a final check.

### CODE REVIEW

File: src/api/fetch.js [hunk 1]
```diff
@@ -1,24 +1,28 @@
...
+import { Scope } from "./scopes";
 const fetchRetry = require("fetch-retry")(fetch, {
   retries: 0,
   retryDelay: function (attempt) {
     return Math.pow(2, attempt) * 15000;
   },
 });
 
-export async function* indexGenerator(path, auth) {
+export async function* indexGenerator(path, auth, scope = new Scope()) {
```
Comment (chvp) [inline src/api/fetch.js:10]: From an API design standpoint, the optionality of this parameter should be clear in the `index` methods, not here.

File: src/api/scopes.js [hunk 2]
```diff
@@ -0,0 +1,72 @@
+export class Scope {
+  scopes = [];
+
+  get finalQuery() {
+    return this.scopes.map((s) => `&${s.key}=${s.query}`).join("");
+  }
+
+  addScope = (key, query) => {
```
Comment (chvp) [inline src/api/scopes.js:8]: Is there a reason you don't write
```suggestion
  addScope(key, query) {
```
?

File: src/api/scopes.js [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+    if (typeof query !== "string" && typeof query !== "number") {
```
Comment (chvp) [inline src/api/scopes.js:9]: This should not be necessary.

File: src/api/scopes.js [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+      throw TypeError(
+        `${query} is an ${typeof query} while your scope '${key}' that expects a string or number.`
+      );
+    }
+    this.scopes.push({ key, query });
+    return this;
+  };
+
+  addScopesFromArray = (key, queries) => {
+    if (!Array.isArray(queries)) {
```
Comment (chvp) [inline src/api/scopes.js:19]: Same here.

File: src/store/actions.js [hunk 5]
```diff
@@ -13,5 +13,8 @@ export async function fetchAll(commit, generator, commitAction) {
     }
   }
   commit(commitAction, results);
-  commit("removeOld");
+  // If a scope is present, we skip removeOld
+  if (scope === null) {
```
Comment (chvp) [inline src/store/actions.js:17]: This does not do what the comment says it does.

File: src/views/albums/Album.vue [hunk 6]
```diff
@@ -61,27 +61,33 @@
...
 
 export default {
   name: "Album",
   components: { AlbumArtists, TracksTable, AlbumActions },
   metaInfo() {
     return { title: this.album?.title };
   },
+  props: {
+    id: {
+      type: [String, Number],
```
Comment (chvp) [inline src/views/albums/Album.vue:78]: The `[String, Number]` feels weird to me. Will this not always be a `String` since it is loaded from the URL? Or does `vue-router` only read the prop from the URL if it was not passed in code?

File: src/views/albums/Album.vue [hunk 7]
```diff
@@ -102,6 +108,24 @@ export default {
...
+    ...mapActions("tracks", { indexTracks: "index" }),
+    async fetchContent(newValue, oldValue) {
+      // After loading the content, the router will change the id from a string to a number
+      // but we don't actually want to load the content twice
+      if (newValue == oldValue) {
+        return;
+      }
+
+      const album = this.read(this.id);
+      const tracks = this.indexTracks(new TracksScope().album(this.id));
```
Comment (chvp) [inline src/views/albums/Album.vue:121]: It feels a little arbitrary what is fetched here. Why not fetch the genres of the tracks, why not fetch the labels?

File: src/views/albums/Album.vue [hunk 8]
```diff
@@ -102,6 +108,24 @@ export default {
...
+      Promise.all([album, tracks]).finally(() => {
```
Comment (chvp) [inline src/views/albums/Album.vue:122]: Why not `return await` this, to make the control flow in the async method explicit?

**Review by chvp — CHANGES_REQUESTED** [review]

File: src/api/fetch.js (hunk 1 above)
Comment (robbevp) [reply src/api/fetch.js:10]: I've also made it clear in the index action where applicable. We also need the optionality here, since not all index actions will provide a scope

File: src/views/albums/Album.vue (hunk 6 above)
Comment (robbevp) [reply src/views/albums/Album.vue:78]: The latter. If you go to `app/albums/1` directly the prop will be a string (and Vue's type checker will complain if that is not an option), if you navigate through the app the prop will be a number.

File: src/api/scopes.js (hunk 2 above)
Comment (robbevp) [reply src/api/scopes.js:8]: Not really, it shouldn't make any difference to write it one way or the other

File: src/views/albums/Album.vue (hunk 7 above)
Comment (robbevp) [reply src/views/albums/Album.vue:121]: I mostly started from the scopes that are available. There isn't (yet) a scope to get genres by tracks or labels by album on the API.

File: src/api/scopes.js (hunk 2 above)
Comment (chvp) [reply src/api/scopes.js:8]: I like the `function() {}` syntax more, tbh.

File: src/views/albums/Album.vue (hunk 6 above)
Comment (chvp) [reply src/views/albums/Album.vue]: OK.

File: src/views/albums/Album.vue (hunk 7 above)
Comment (chvp) [reply src/views/albums/Album.vue:121]: Ah, makes sense.

File: src/views/albums/Album.vue [hunk 9]
```diff
@@ -61,27 +61,33 @@
...
 
 export default {
   name: "Album",
   components: { AlbumArtists, TracksTable, AlbumActions },
   metaInfo() {
     return { title: this.album?.title };
   },
+  props: {
+    id: {
+      type: [String],
```
Comment (chvp) [inline src/views/albums/Album.vue:78]: Why did you remove the `, Number` here?

Comment (robbevp) [reply src/views/albums/Album.vue]: I was experimenting with these types to reconstruct why I added both. This slipped in by accident.

**Review by chvp — APPROVED** [review]
